### PR TITLE
boards: arm: Add Renesas Spider board

### PR DESCRIPTION
**Pull request by aaillet**
This pull request is introducing Generation 4 of Renesas R-Car SoCs with the S4 SoC as well as the Spider board into Zephyr.

Gen4 SoCs vs Gen3 (Zephyr POV) :

Cortex-R52 is now used as Real Time core
We are now able to use embedded ARM timer
Some IPs are the same and will use existing drivers, sometime with few improvements (UART, GPIO...)
Introduction of new IPs
S4 SoC that is equipping multiple Gen4 boards have some Cortex-A55s and a Cortex-R52 core.

Spider board is the reference board for S4 SoC, developpment oriented, this board is providing a lot of available IOs.

Operation of basic samples such as hello_world and blinky have been validated.

### CODE REVIEW

**Comment by aaillet** [comment]
This PR is ready to be reviewed as th last CI failures is a false positive !
`pinctrl_soc_pin_t` has to be defined as documented here : https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html#pin-control-drivers.

**Comment by MaureenHelm** [comment]
@gmarull @povergoing  can you take a look please?

**Review by povergoing — DISMISSED** [review]
LGTM for the R52 part

File: soc/arm/renesas_rcar/common/pinctrl_rcar.h [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ * @brief set the register index for a given pin
+ *
+ * @param the pin
+ * @param pointer for the resulting register index
+ * @return 0 if the register index is found, negative
+ * errno otherwise.
+ */
+int pfc_rcar_get_reg_index(uint8_t pin, uint8_t *reg_index);
```
Comment (gmarull) [inline soc/arm/renesas_rcar/common/pinctrl_rcar.h:117]: this should not be part of pinctrl_soc, as it's something needed internally by drivers. please, create a new internal header so that drivers/pinctrl.h is not polluted with unnecessary stuff.

**Review by gmarull — DISMISSED** [review]

**Review by henrikbrixandersen — COMMENTED** [review]
GPIO changes okay.

File: soc/arm/renesas_rcar/common/pinctrl_rcar.h (hunk 1 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/common/pinctrl_rcar.h:117]: Hi @gmarull ,
Thanks for your comment.
If my understanding is fine, you want me to move these common declaration to a new file here `zephyr/include/zephyr/drivers/pinctrl/pinctrl_rcar_common.h` ?

File: soc/arm/renesas_rcar/common/pinctrl_rcar.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	DT_PROP(node_id, bias_pull_up)   * RCAR_PIN_PULL_UP |		       \
+	DT_PROP(node_id, bias_pull_down) * RCAR_PIN_PULL_DOWN |		       \
+	DT_PROP(node_id, bias_disable)   * RCAR_PIN_PULL_DISABLE |	       \
+	RCAR_HAS_IPSR(node_id) * RCAR_PIN_FLAGS_FUNC_SET
+
+#define RCAR_DT_PIN(node_id)						       \
+	{								       \
+		.pin = DT_PROP_BY_IDX(node_id, pin, 0),			       \
+		.func = COND_CODE_1(RCAR_HAS_IPSR(node_id),		       \
+				    (RCAR_PIN_FUNC(node_id)), (0)),	       \
```
Comment (xakep-amatop) [inline soc/arm/renesas_rcar/common/pinctrl_rcar.h:63]: pls, use struct initializer `{0}` for this field

File: soc/arm/renesas_rcar/common/pinctrl_rcar.h (hunk 1 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/common/pinctrl_rcar.h:117]: Hi @gmarull, I'm currently reworking this PR to merge some changes with #60252 , can you confirm that I understood well your remark before I proceed ? Thanks

**Comment by aaillet** [comment]
Just updated the PR with the following changes:
- Rebased on main
- Removed NMI init in soc.c following #59869
- Moved GIC version to DT following #58035
- Updated clock driver following #57172 thanks @xakep-amatop for the help
- Updated pinctrl implementation for Gen4 SoCs

**Review by xakep-amatop — CHANGES_REQUESTED** [review]

File: drivers/pinctrl/pfc_rcar.c [hunk 3]
```diff
@@ -217,10 +252,13 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
 	return ret;
 }
 
+#if defined(CONFIG_SOC_SERIES_RCAR_GEN3)
 __boot_func static int pfc_rcar_driver_init(void)
 {
 	DEVICE_MMIO_TOPLEVEL_MAP(pfc, K_MEM_CACHE_NONE);
+	reg_base[0] = DEVICE_MMIO_TOPLEVEL_GET(pfc);
```
[the remark this answers is not in the document]
Comment (aaillet) [reply drivers/pinctrl/pfc_rcar.c:259]: thanks, fixed !

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+static int r8a779f0_set_rate_helper(uint32_t module, uint32_t *divider, uint32_t *div_mask)
+{
+	int ret = -ENOTSUP;
+
+	switch (module) {
+	default:
+		break;
+	}
```
Comment (carlocaione) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:151]: why?

File: soc/arm/renesas_rcar/gen4/soc.c [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+void z_arm_platform_init(void)
+{
+	L1C_DisableCaches();
+	L1C_DisableBTAC();
+
+	/* Invalidate instruction cache and flush branch target cache */
+	__set_ICIALLU(0);
+	__DSB();
+	__ISB();
```
Comment (carlocaione) [inline soc/arm/renesas_rcar/gen4/soc.c:19]: use the barriers API, see `include/zephyr/sys/barrier.h`

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	int ret = 0;
+	uint32_t reg;
+
+	enable = !!enable;
+
+	switch (clk_info->module) {
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
```
Comment (carlocaione) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:63]: this is odd.

**Review by carlocaione — DISMISSED** [review]

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c (hunk 4 above)
Comment (aaillet) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:151]: At the moment, none of the clock used by arm32 support has to be enable here !
The driver is however templated on r8a7795 driver and prepared for arm64 support comming directly after this PR (see #60252)

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c (hunk 6 above)
Comment (aaillet) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:63]: same as for last question, for now, not any clocks used by arm32 support has to be enable manually.
however, this will be used as soon as arm64 support will come up.
These methods are alos needed to be filled into `rcar_cpg_mssr_data`
Would it be enough to add a comment in the driver explaining that waiting for this list to be filled ?

Comment (carlocaione) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:63]: yeah, otherwise this is going to confuse a lot of reviewers :smile:

File: soc/arm/renesas_rcar/gen4/soc.c (hunk 5 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/gen4/soc.c]: Thanks, fixed

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c (hunk 6 above)
Comment (aaillet) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c]: proposed a fix, commenting these part of code, let me know if it can be enough, waiting for arm64 side PR (adding some clocks to these switch case) to be merged !

File: drivers/pinctrl/pfc_rcar.c [hunk 7]
```diff
@@ -10,13 +10,24 @@
...
-#define PFC_RCAR_GPSR 0x100
-#define PFC_RCAR_IPSR 0x200
+
+/* Gen3 only has one base address, Gen4 has one per GPIO controller */
+#if defined(CONFIG_SOC_SERIES_RCAR_GEN3)
+static const uint32_t reg_base[] = {DT_INST_REG_ADDR(0)};
+#elif defined(CONFIG_SOC_SERIES_RCAR_GEN4)
+/* swap both arguments */
+#define PFC_REG_ADDRESS(idx, node_id) DT_REG_ADDR_BY_IDX(node_id, idx)
+static const uint32_t reg_base[] = {
```
[the remark this answers is not in the document]
Comment (aaillet) [reply drivers/pinctrl/pfc_rcar.c]: Thanks for your comment, fixed

File: soc/arm/renesas_rcar/common/pinctrl_rcar.h (hunk 2 above)
Comment (aaillet) [reply soc/arm/renesas_rcar/common/pinctrl_rcar.h]: Fixed :+1:

File: drivers/pinctrl/pfc_rcar.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaillet) [reply drivers/pinctrl/pfc_rcar.c]: Fixed, thanks

**Review by xakep-amatop — DISMISSED** [review]

File: soc/arm/renesas_rcar/common/pinctrl_rcar.h (hunk 1 above)
Comment (gmarull) [reply soc/arm/renesas_rcar/common/pinctrl_rcar.h:118]: What I mean is that this should only be exposed to code requiring it. An internal header is fine, but decoupled from pinctrl_soc.h

Comment (aaillet) [reply soc/arm/renesas_rcar/common/pinctrl_rcar.h]: Fixed it, please tel me if it's ok now !

**Comment by aaillet** [comment]
Updated the PR moving pinctrl private function prototype to a new internal header.

File: drivers/pinctrl/pfc_rcar.h [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * @brief set the register index for a given pin
+ *
+ * @param the pin
+ * @param pointer for the resulting register index
+ * @return 0 if the register index is found, negative
+ * errno otherwise.
+ */
+int pfc_rcar_get_reg_index(uint8_t pin, uint8_t *reg_index);
```
Comment (gmarull) [inline drivers/pinctrl/pfc_rcar.h:21]: this header is not self-contained

Comment (aaillet) [reply drivers/pinctrl/pfc_rcar.h:21]: Sorry, I was not aware of this rule

So, I moved the "pfc_rcar.h" include to the top of pfc_rcar.c includes as proposed in #41543 , it also allows me to test self-containment of this new header.

For it to be fully self-contained, I think I need to include pinctrl.soc that is defining both `pfc_bias_reg` and `pfc_drive_reg`  struct.
I'm asking this because, including stdint is enough for it to build (it seems like gcc is not checking struct def at build) but it seems wrong to me.

Researches led me to this paper[1] explaining that all "incomplete" types would be managed the same at build.

What would be Zephyr guideline in this case ?

[1] : https://beej.us/guide/bgc/html/split/incomplete-types.html

**Comment by klogg** [comment]
Hi @aaillet
Do you plan to push this? Seems stalled...

**Comment by aaillet** [comment]
> Hi @aaillet Do you plan to push this? Seems stalled...

Yes I plan to.
Not stalled, waiting for reviews, I will tag it to be talked about in a dev meeting soon.

**Comment by aaillet** [comment]
- Rebased on main
- Added gpio key field to board dts as introduced in #61115 
Trying to fix CI

**Review by xakep-amatop — DISMISSED** [review]

**Comment by aaillet** [comment]
- Rebased on main
- Disabled L1 cache (introducing issues with reading registers values being cached)

**Comment by xakep-amatop** [comment]
> * Rebased on main

Note for the future: the Zephyr documentation recommends performing a push rebase and making your changes in separate updates [Workflow suggestions that help reviewers](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers). 

>     * Disabled L1 cache (introducing issues with reading registers values being cached)

Can you share some details about where you performed disabling and which registers you encountered problems with?

**Comment by aaillet** [comment]
> > * Rebased on main
> 
> Note for the future: the Zephyr documentation recommends performing a push rebase and making your changes in separate updates [Workflow suggestions that help reviewers](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers).
> 
Thanks for the link, I'll do it that way next time :+1: 
> > ```
> > * Disabled L1 cache (introducing issues with reading registers values being cached)
> > ```
> 
> Can you share some details about where you performed disabling and which registers you encountered problems with?

Here is how we activated cache for gen4 at first (same way it it done for gen3 socs) : https://github.com/iotbzh/zephyr/commit/96a79f89ae831d56747c09f735a75b4daf26b6df

We encountered problem while working on IPC driver. 
Zephyr was not able to read updated master status through libmetal (it was confirmed that ra.
Even if linux was updating the value of this shared register, Zephyr was still reading the old value.
It was fixed instantly by disabling cache.

I removed it of this PR to avoid contagion to other drivers we might not have seen.
We will soon investigate on this topic

**Review by xakep-amatop — DISMISSED** [review]

**Comment by aaillet** [comment]
Removed a comment in board dts (was not respecting compliance), was useless.

**Review by xakep-amatop — DISMISSED** [review]

**Comment by aaillet** [comment]
@lorc @gmarull @povergoing can you please take another look ?

File: dts/arm/renesas/rcar/gen4/r8a779f0.dtsi [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+			status = "disabled";
+		};
+
+		/* Zephyr console */
+		scif0: serial@e6e60000 {
+			interrupts = <GIC_SPI 249 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+			clocks = <&cpg CPG_MOD 702>, <&cpg CPG_CORE R8A779F0_CLK_S0D12_PER>;
+		};
+
+		/* Linux console */
```
[the remark this answers is not in the document]
Comment (lorc) [reply dts/arm/renesas/rcar/gen4/r8a779f0.dtsi:64]: AFAIK, Cortex A cluster uses HSCIF0 nowadays.

**Review by lorc — COMMENTED** [review]
I have a small comment regarding SCIF3/HSCIF0 used by Cortex A cluster. If I remember correctly, Renesas BSP moved Cortex A to HSCIF0. At least this is what we use in OP-TEE, U-Boot and Xen.

**Comment by aaillet** [comment]
> I have a small comment regarding SCIF3/HSCIF0 used by Cortex A cluster. If I remember correctly, Renesas BSP moved Cortex A to HSCIF0. At least this is what we use in OP-TEE, U-Boot and Xen.

I think that we agree on this subject.

Physically SCIF3 and HSCIF0 are using the same lines and Cortex A's are using HSCIF0. (HSCIF1 and SCIF0 are on the same line as well)
Zephyr part is only supporting SCIF so I listed physically available lines in the device tree.
As typical use case is that Cortex A's are using HSCIF0, I selected SCIF0 as main Zephyr serial port.

For other use cases, Zephyr may use SCIF3 line.

Please keep in mind that HSCIF0 and SCIF0 are not on the same line, they can work at the same without any issues. (On different physical port)

I hope I answered your comments.

**Review by lorc — DISMISSED** [review]
Yes, thank you.

File: boards/arm/rcar_spider/doc/rcar_spider.rst [hunk 10]
```diff
@@ -0,0 +1,187 @@
+.. _rcar_spider_boards:
+
+Renesas R-Car Spider
+####################
+
+Overview
+********
+
+R-Car S4 enables the launch of Car Server/CoGW with high performance, high-speed networking, high security and high functional safety levels that are required as E/E architectures evolve into domains and zones.
```
Comment (fabiobaltieri) [inline boards/arm/rcar_spider/doc/rcar_spider.rst:9]: can you wrap the lines to either 80 or 100 columns? sorry we don't have a check for this yet but that's how these docs are normally done

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+};
+
+static int r8a779f0_cpg_enable_disable_core(const struct device *dev,
+					   struct cpg_clk_info_table *clk_info,
+					   uint32_t enable)
+{
+	int ret = -ENOTSUP;
+	uint32_t reg;
+
+	enable = !!enable;
```
Comment (fabiobaltieri) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:57]: does not look like this is used, stale? guess you  may want to remove the argument entirely

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+static int r8a779f0_cpg_core_clock_endisable(const struct device *dev,
+					    struct rcar_cpg_clk *clk,
+					    bool enable)
+{
+	struct cpg_clk_info_table *clk_info;
+	struct r8a779f0_cpg_mssr_data *data = dev->data;
+	k_spinlock_key_t key;
+	int ret = 0;
```
Comment (fabiobaltieri) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:78]: drop the initializer, return with `return 0;` at the end of the function

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+	k_spin_unlock(&data->cmn.lock, key);
+
+	return ret;
+}
+
+int r8a779f0_cpg_mssr_start_stop(const struct device *dev,
+	clock_control_subsys_t sys, bool enable)
+{
+	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *)sys;
+	int ret = -EINVAL;
```
Comment (fabiobaltieri) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:108]: no need to initialize this

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+	} else if (clk->domain == CPG_CORE) {
+		ret = r8a779f0_cpg_core_clock_endisable(dev, clk, enable);
+	}
+
+	return ret;
+}
+
+static uint32_t r8a779f0_get_div_helper(uint32_t reg_val, uint32_t module)
+{
+	uint32_t divider = RCAR_CPG_NONE;
```
Comment (fabiobaltieri) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:130]: how about using explicit return and ditching the variable? easier to follow...

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+	default:
+		break;
+	}
+
+	return divider;
+}
+
+static int r8a779f0_set_rate_helper(uint32_t module, uint32_t *divider, uint32_t *div_mask)
+{
+	int ret = -ENOTSUP;
```
Comment (fabiobaltieri) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:146]: just `return -ENOSUP` for now, you can add stuff later with the structure that makes sense at that point

File: boards/arm/rcar_spider/doc/rcar_spider.rst (hunk 10 above)
Comment (aaillet) [reply boards/arm/rcar_spider/doc/rcar_spider.rst:9]: Ok, but how should I manage links that are longer than 100 characters then ?
They are gathered at the bottom of the rst file

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c (hunk 11 above)
Comment (aaillet) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:57]: It's not used yet as current driver don't need their clocks to be enabled or disabled manually.
This whole function is a placeholder to keep skeleton from other renesas clock driver.
It's also already planned that the arm64 support of this soc (coming just after this PR, see #60252 ) will need some clocks to be enable/disabled for some drivers.

More specifically, this argument is just as a boolean to know if this function is called to enable or disable the concerned clock.

That's all reasons why this function has to be here but is not used yet.

File: boards/arm/rcar_spider/doc/rcar_spider.rst (hunk 10 above)
Comment (fabiobaltieri) [reply boards/arm/rcar_spider/doc/rcar_spider.rst:9]: those are fine, just wrap the normal text

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c (hunk 11 above)
Comment (fabiobaltieri) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:57]: ok, keep the argument, drop this line? or mark it as `ARG_UNUSED`

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c (hunk 13 above)
Comment (aaillet) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:108]: was getting "may be used uninitialized" warning when declaring ret uninitialized

adding an "else" initializing ret at -EINVAL to fix warning

**Comment by aaillet** [comment]
Addressed comments from @fabiobaltieri concerning board documentation and clock driver.

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+	key = k_spin_lock(&data->cmn.lock);
+	r8a779f0_cpg_enable_disable_core(dev, clk_info, enable);
+	k_spin_unlock(&data->cmn.lock, key);
+
+	return 0;
+}
+
+int r8a779f0_cpg_mssr_start_stop(const struct device *dev,
+	clock_control_subsys_t sys, bool enable)
```
Comment (fabiobaltieri) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:106]: sorry one more indentation nitpick: under the previous parenthesis is how it's normally done here, you are mixing thigs up a bit in the file, here you are just doing a tab, in `r8a779f0_cpg_mssr_start` you are back one charcter, could you check these and make sure they are at least coherent within the file?

```
int r8a779f0_cpg_mssr_start_stop(const struct device *dev,
                                 clock_control_subsys_t sys, bool enable)
static int r8a779f0_cpg_mssr_start(const struct device *dev,
                                   clock_control_subsys_t sys)
```

is how it should be

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+	switch (clk_info->module) {
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
+
+	if (!ret) {
+		rcar_cpg_write(DEVICE_MMIO_GET(dev), clk_info->offset, reg);
+	}
+	return ret;
```
Comment (fabiobaltieri) [inline drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:69]: would not you be better off just `return -ENOTUP` and then implement the whole thing once there's something to do? I understand you want to have the stubs but it just looks really really weird right now, plus the way you are using ret around is hard to follow and does not look quite right

Comment (aaillet) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:69]: I just need to check if it is mandatory to write clk reg when errors are happening (l67) in the controller doc.
If it is, I'll just do this `rcar_cpg_write` before returning ENOTSUP :+1: 

Yeah, I totally understand that these placeholders are weird right now, at least keeping them with a `return -ENOTSUP` is already nice for the driver creation !

File: drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c (hunk 16 above)
Comment (aaillet) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:106]: running clang-format on it as it is a new file could be the right call ?

Comment (fabiobaltieri) [reply drivers/clock_control/clock_control_r8a779f0_cpg_mssr.c:106]: clang-format mostly works but often does a bit of a disaster on complex macros, if you want to use it fine but don't take the output religiously, either commit locally, clang-format and then manually `checkout -p` the messy chunks, or add some `/* clang-format off */` `/* clang-format on */` markers around the part it messes up, if any.

**Comment by aaillet** [comment]
Addressed last comments from @fabiobaltieri concerning clock driver

**Review by fabiobaltieri — APPROVED** [review]

**Review by lorc — APPROVED** [review]

**Review by xakep-amatop — APPROVED** [review]
